File: src/main/java/com/openlattice/ids/IdCatchupEntryProcessor.java
/*
 * Copyright (C) 2017. OpenLattice, Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * You can contact the owner of the copyright at support@openlattice.com
 *
 */

package com.openlattice.ids;

import static com.google.common.base.Preconditions.checkNotNull;

import com.kryptnostic.rhizome.hazelcast.processors.AbstractRhizomeEntryProcessor;
import com.openlattice.postgres.PostgresColumn;
import com.openlattice.postgres.PostgresTable;
import com.openlattice.postgres.ResultSetAdapters;
import com.zaxxer.hikari.HikariDataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Map.Entry;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
public class IdCatchupEntryProcessor extends AbstractRhizomeEntryProcessor<Integer, Range, Void> {
    private static final Logger logger = LoggerFactory.getLogger( IdCatchupEntryProcessor.class );

    private final HikariDataSource hds;

    public IdCatchupEntryProcessor( HikariDataSource hds ) {
        this.hds = checkNotNull( hds );
    }

    @Override
    public Void process( Entry<Integer, Range> entry ) {
        final Range range = checkNotNull( entry.getValue() ); //Range should never be null in the EP.
        int counter = 0;
        try ( final Connection connection = hds.getConnection(); final var ps = prepareExistQuery( connection ) ) {

            while ( exists( ps, range.peek() ) ) {
                range.nextId();
                counter++;
                if ( ( counter % 10000 ) == 0 ) {
                    logger.info( "Incremented range with base {} by {}", range.getBase(), counter );
                }
            }

            entry.setValue( range );
            logger.info( "Caught up range with base {} by {} increments", range.getBase(), counter );
        } catch ( SQLException e ) {
            logger.error( "Error catching up ranges.", e );
        }
        return null;
    }

    public PreparedStatement prepareExistQuery( Connection connection ) throws SQLException {
        return connection.prepareStatement(
                "SELECT count(*) from " + PostgresTable.IDS.getName() +
                        " WHERE " + PostgresColumn.ID_VALUE.getName() + " = ?" );
    }

    public boolean exists( PreparedStatement ps, UUID id ) throws SQLException {
        ps.setObject( 1, id );
        //Count query always guaranteed to have one row.
        final var rs = ps.executeQuery();
        rs.next();
        return ResultSetAdapters.count( rs ) > 0;
    }
}

File: src/main/kotlin/com/openlattice/ids/IdCatchupEntryProcessor.kt
package com.openlattice.ids


import com.google.common.base.Preconditions.checkNotNull
import com.kryptnostic.rhizome.hazelcast.processors.AbstractRhizomeEntryProcessor
import com.openlattice.postgres.PostgresColumn
import com.openlattice.postgres.PostgresTable
import com.openlattice.postgres.ResultSetAdapters
import com.openlattice.postgres.ResultSetAdapters.id
import com.zaxxer.hikari.HikariDataSource
import org.slf4j.LoggerFactory
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.SQLException

/**
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
class IdCatchupEntryProcessor(hds: HikariDataSource) : AbstractRhizomeEntryProcessor<Int, Range?, Void>() {

    private val hds: HikariDataSource = checkNotNull(hds)

    companion object {

        private val logger = LoggerFactory.getLogger(IdCatchupEntryProcessor::class.java)
    }

    override fun process(entry: MutableMap.MutableEntry<Int, Range?>): Void? {
        val range = entry.value!! //Range should never be null in the EP.
        val lowerbound = Range(range.base).nextId()
        val supremum = Range(range.base + (1 shl 48)).nextId()

        try {
            hds.connection.use { connection ->
                prepareLatestQuery(connection).use { ps ->
                    ps.setObject(1, lowerbound)
                    ps.setObject(2, supremum)
                    val rs = ps.executeQuery()
                    val newMaxRange = if (rs.next()) {
                        val id = id(rs)
                        val r = Range(
                                range.base,
                                id.mostSignificantBits xor range.base,
                                id.leastSignificantBits
                        )
                        r.nextId()
                        r
                    } else {
                        logger.warn("Detected empty range with base {} ({})", range.base, range.base ushr 48)
                        Range(range.base) //This is -2 so that increment takes it to -1
                    }
                    entry.setValue(newMaxRange)
                }
            }
        } catch (e: SQLException) {
            logger.error("Error catching up ranges.", e)
        }

        return null
    }

    /**
     * This only works because the base represents the 16 most significant bits in the UUID. This query is intended to
     * be used with the lowest possible UUID for a given base and the lowest possible UUID for the immediately larger
     * base.
     *
     * @param connection A jdbc connection to a database with a valid [PostgresTable.IDS] table.
     * @return A prepared statement that returns the largest UUID in a given base. As long as preceeding assumptions are
     * met.
     */
    @Throws(SQLException::class)
    private fun prepareLatestQuery(connection: Connection): PreparedStatement {
        return connection.prepareStatement(
                "SELECT ${PostgresColumn.ID.name} from ${PostgresTable.IDS.name} WHERE ${PostgresColumn.ID_VALUE.name} >= ? AND ${PostgresColumn.ID_VALUE.name} < ? ORDER BY ID DESC LIMIT 1"
        )
    }

}

